src/Workspaces/Core/Portable/Workspace/ImmutableWorkspace.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis;
+
+internal sealed class ImmutableWorkspace : Workspace
+{
+    public ImmutableWorkspace(HostServices host, string? workspaceKind)
+        : base(host, workspaceKind)
+    {
+    }
+
+    internal override bool CanAddProjectReference(ProjectId referencingProject, ProjectId referencedProject) => false;
+    public override bool CanApplyChange(ApplyChangesKind feature) => false;
+    protected override bool CanApplyCompilationOptionChange(CompilationOptions oldOptions, CompilationOptions newOptions, Project project) => false;
+    public override bool CanApplyParseOptionChange(ParseOptions oldOptions, ParseOptions newOptions, Project project) => false;
+    internal override bool CanChangeActiveContextDocument => false;
+    public override bool CanOpenDocuments => false;
+    internal override bool CanUpdateOptions => false;
```

Comment:
consider overriding TryApplyChanges to be a no-op/throw.

Reply:
no override of CanUpdateSolution?  Intentional?  I'll admit it's not quite clear what semantics we want here.

Reply:
Forgot to remove CanUpdateSolution. Not needed anymore.